src/day01.ml

```diff
@@ -1,2 +1,123 @@
-let () =
-  print_endline "hello"
+type turn_dir =
+  | Left
+  | Right
+
+type movement = turn_dir * int
+
+module Parser = struct
+  open Angstrom
+
+  let turn_left = char 'L' *> return Left
+  let turn_right = char 'R' *> return Right
+
+  let integer =
+    take_while1 (function '0' .. '9' -> true | _ -> false) >>| int_of_string
+
+  let instruction =
+    (turn_left <|> turn_right) >>= (fun d ->
+        integer >>| fun i -> (d, i))
+
+  let sep = char ',' *> skip_many (char ' ')
+
+  let instructions = sep_by sep instruction
+
+  let parse_input inputstr =
+    match parse_only instructions (`String inputstr) with
+    | Result.Ok v -> v
+    | Result.Error msg -> failwith msg
+end
+
+
+module Args = struct
+  open Cmdliner
+
+  let input_filename =
+    let doc = "file containing input data" in
+    Arg.(required & pos 0 (some file) None & info [] ~docv:"FILENAME" ~doc)
+
+  let part =
+    let doc = "part 1 or part 2" in
+    Arg.(required & pos 1 (some int) None & info [] ~docv:"PART" ~doc)
+
+  let run_main mainfunc =
+    let main_t = Term.(const mainfunc $ input_filename $ part) in
+    let info = Term.info "day01" ~doc:"day 1 solver" in
+    match Term.eval (main_t, info) with `Error _ -> exit 1 | _ -> exit 0
+end
+
+type dir = N | E | S | W
+
+let turn dir turn_dir =
+  match (dir, turn_dir) with
+  | N, Left | S, Right -> W
+  | E, Left | W, Right -> N
+  | S, Left | N, Right -> E
+  | W, Left | E, Right -> S
+
+let walk_block (n, e, d) distance = match d with
+  | N -> (n + distance, e, N)
+  | E -> (n, e + distance, E)
+  | S -> (n - distance, e, S)
+  | W -> (n, e - distance, W)
+
+let apply_instruction (n, e, d) (turn_dir, distance) =
+  walk_block (n, e, (turn d turn_dir)) distance
+
+
+let main_1 walk =
+  List.fold_left apply_instruction (0, 0, N) walk
+
+
+module IntPairs =
+struct
+  type t = int * int
+  let compare (x0,y0) (x1,y1) =
```

Comment:
If I tested this right in the Toplevel, the default comparison behaviour on tuples is exactly this.